### ISSUE TITLE
sovits微调时，偶尔出现NAN

### ISSUE THREAD

**Issue by adele-php** [RVC-Boss/GPT-SoVITS#1156]
![1RKD_D)5%G7EBA29~3MC8~T](https://github.com/RVC-Boss/GPT-SoVITS/assets/29390522/5beb66ab-8b63-452d-b1cb-4b9560f1bf7c)

代码用的时2024.05.31 clone下来的代码
显卡和环境如下
centos7.8
![image](https://github.com/RVC-Boss/GPT-SoVITS/assets/29390522/e0d6db52-579d-49b6-988c-6bc7e6a456e8)

**Comment by adele-php** [comment]
今天把代码重新克隆了一下，然后微调sovits还是出现NAN，训练日志如下

/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/optim/lr_scheduler.py:136: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/utils/data/dataloader.py:557: UserWarning: This DataLoader will create 6 worker processes in total. Our suggested max number of worker in current system is 4, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
0it [00:00, ?it/s]/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: stft with return_complex=False is deprecated. In a future pytorch release, stft will return complex tensors for all inputs, and return_complex=False will raise an error.
Note: you can still call torch.view_as_real on the complex output to recover the old return format. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/native/SpectralOps.cpp:863.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: stft with return_complex=False is deprecated. In a future pytorch release, stft will return complex tensors for all inputs, and return_complex=False will raise an error.
Note: you can still call torch.view_as_real on the complex output to recover the old return format. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/native/SpectralOps.cpp:863.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: stft with return_complex=False is deprecated. In a future pytorch release, stft will return complex tensors for all inputs, and return_complex=False will raise an error.
Note: you can still call torch.view_as_real on the complex output to recover the old return format. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/native/SpectralOps.cpp:863.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: stft with return_complex=False is deprecated. In a future pytorch release, stft will return complex tensors for all inputs, and return_complex=False will raise an error.
Note: you can still call torch.view_as_real on the complex output to recover the old return format. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/native/SpectralOps.cpp:863.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: stft with return_complex=False is deprecated. In a future pytorch release, stft will return complex tensors for all inputs, and return_complex=False will raise an error.
Note: you can still call torch.view_as_real on the complex output to recover the old return format. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/native/SpectralOps.cpp:863.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: stft with return_complex=False is deprecated. In a future pytorch release, stft will return complex tensors for all inputs, and return_complex=False will raise an error.
Note: you can still call torch.view_as_real on the complex output to recover the old return format. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/native/SpectralOps.cpp:863.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: stft with return_complex=False is deprecated. In a future pytorch release, stft will return complex tensors for all inputs, and return_complex=False will raise an error.
Note: you can still call torch.view_as_real on the complex output to recover the old return format. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/native/SpectralOps.cpp:863.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/functional.py:650: UserWarning: ComplexHalf support is experimental and many operators don't support it yet. (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/aten/src/ATen/EmptyTensor.cpp:31.)
  return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
[W reducer.cpp:1346] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters in the forward pass. This flag results in an extra traversal of the autograd graph every iteration,  which can adversely affect performance. If your model indeed never has any unused parameters in the forward pass, consider turning this flag off. Note that this warning may be a false positive if your model has flow control causing later iterations to have unused parameters. (function operator())
/anaconda3/envs/GPTSoVits3/lib/python3.9/site-packages/torch/autograd/__init__.py:251: UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
grad.sizes() = [1, 9, 96], strides() = [55584, 96, 1]
bucket_view.sizes() = [1, 9, 96], strides() = [864, 96, 1] (Triggered internally at /opt/conda/conda-bld/pytorch_1699449181081/work/torch/csrc/distributed/c10d/reducer.cpp:320.)
  Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
INFO:0605-4xiejingguididufeipang:Train Epoch: 1 [0%]
INFO:0605-4xiejingguididufeipang:[2.7567014694213867, 1.9121623039245605, 9.378927230834961, 20.815147399902344, 0.0, 2.5181612968444824, 0, 9.99875e-05]
42it [00:40,  1.03it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 1
42it [00:33,  1.27it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 2
16it [00:14,  1.42it/s]INFO:0605-4xiejingguididufeipang:Train Epoch: 3 [38%]
INFO:0605-4xiejingguididufeipang:[2.56430721282959, 2.0105438232421875, 9.521171569824219, 18.557138442993164, 0.0, 2.4015772342681885, 100, 9.996250468730469e-05]
42it [00:33,  1.25it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 3
42it [00:32,  1.29it/s]
INFO:0605-4xiejingguididufeipang:Saving model and optimizer state at iteration 4 to logs/0605-4xiejingguididufeipang/logs_s2/G_233333333333.pth
INFO:0605-4xiejingguididufeipang:Saving model and optimizer state at iteration 4 to logs/0605-4xiejingguididufeipang/logs_s2/D_233333333333.pth
INFO:0605-4xiejingguididufeipang:saving ckpt 0605-4xiejingguididufeipang_e4:Success.
INFO:0605-4xiejingguididufeipang:====> Epoch: 4
32it [00:25,  1.46it/s]INFO:0605-4xiejingguididufeipang:Train Epoch: 5 [76%]
INFO:0605-4xiejingguididufeipang:[nan, nan, nan, nan, 0.0, nan, 200, 9.993751562304699e-05]
42it [00:33,  1.26it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 5
42it [00:32,  1.30it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 6
42it [00:32,  1.30it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 7
6it [00:07,  1.08s/it]INFO:0605-4xiejingguididufeipang:Train Epoch: 8 [14%]
INFO:0605-4xiejingguididufeipang:[nan, nan, nan, nan, 0.0, nan, 300, 9.990004373906418e-05]
42it [00:33,  1.26it/s]
INFO:0605-4xiejingguididufeipang:Saving model and optimizer state at iteration 8 to logs/0605-4xiejingguididufeipang/logs_s2/G_233333333333.pth
INFO:0605-4xiejingguididufeipang:Saving model and optimizer state at iteration 8 to logs/0605-4xiejingguididufeipang/logs_s2/D_233333333333.pth
INFO:0605-4xiejingguididufeipang:saving ckpt 0605-4xiejingguididufeipang_e8:Success.
INFO:0605-4xiejingguididufeipang:====> Epoch: 8
42it [00:32,  1.30it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 9
22it [00:18,  1.47it/s]INFO:0605-4xiejingguididufeipang:Train Epoch: 10 [52%]
INFO:0605-4xiejingguididufeipang:[2.5033650398254395, 2.068312406539917, 7.779954433441162, 19.149423599243164, 0.0, 34.55687713623047, 400, 9.987507028906759e-05]
42it [00:33,  1.27it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 10
42it [00:32,  1.30it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 11
38it [00:29,  1.46it/s]INFO:0605-4xiejingguididufeipang:Train Epoch: 12 [90%]
INFO:0605-4xiejingguididufeipang:[nan, nan, nan, nan, 0.0, nan, 500, 9.98501030820433e-05]
42it [00:32,  1.29it/s]
INFO:0605-4xiejingguididufeipang:Saving model and optimizer state at iteration 12 to logs/0605-4xiejingguididufeipang/logs_s2/G_233333333333.pth
INFO:0605-4xiejingguididufeipang:Saving model and optimizer state at iteration 12 to logs/0605-4xiejingguididufeipang/logs_s2/D_233333333333.pth
INFO:0605-4xiejingguididufeipang:saving ckpt 0605-4xiejingguididufeipang_e12:Success.
INFO:0605-4xiejingguididufeipang:====> Epoch: 12
42it [00:32,  1.29it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 13
42it [00:32,  1.30it/s]
INFO:0605-4xiejingguididufeipang:====> Epoch: 14
12it [00:11,  1.37it/s]INFO:0605-4xiejingguididufeipang:Train Epoch: 15 [29%]
INFO:0605-4xiejingguididufeipang:[nan, nan, nan, 20.303781509399414, 0.0, 31.83603286743164, 600, 9.981266397366609e-05]
42it [00:31,  1.34it/s]

**Comment by Separatee** [comment]
据我推断，这是由于:
过高的学习率会**导致梯度爆炸**，从而使得模型参数变为nan。
请尝试降低一个数量级的学习率(缩一个小数点位。)
> (数据集很多的时候，降低学习率是非常有必要的)。
数据预处理问题：输入数据可能包含无效值或极端值，需要确保数据清洗和预处理的正确性，检查你的音频denoise_opt是否正确一键三连是否出错。必要时一个一个手动执行。
还有可能是切分音频问题，检查slicer_opt文件夹内音频是否正常。时间略长导致梯度爆炸，无法正确识别token

**Comment by RVC-Boss** [comment]
你的显卡型号是什么？（猜想也有可能和显卡有关）

**Comment by adele-php** [comment]
显卡型号上面的图里有，是tesla T4

**Comment by RVC-Boss** [comment]
> 显卡型号上面的图里有，是tesla T4

试试fp32训练。configs/s2.json里有个fp16_run False改True

ps：我训练sovits底膜也训出过很大的loss，改成fp32就好了
